build/xtask/src/flash.rs

```
@@ -11,59 +12,233 @@ use anyhow::Context;
+}
+
+impl FlashProgramConfig {
+    fn new(path: Vec<&str>) -> Self {
+        FlashProgramConfig::Path(path.iter().map(|f| f.to_string()).collect())
+    }
+}
+
+impl FlashConfig {
+    fn new(program: FlashProgram) -> Self {
+        FlashConfig {
+            program: program,
+            args: vec![],
+        }
+    }
+
+    //
+    // Add a command-line argument to the flash program
+    //
+    fn arg<'a>(&'a mut self, val: &str) -> &'a mut Self {
+        self.args.push(FlashArgument::Direct(val.to_string()));
+        self
+    }
+
+    //
+    // Add the path to the payload as an argument to the flash program
+    //
+    fn payload<'a>(&'a mut self) -> &'a mut Self {
+        self.args.push(FlashArgument::Payload);
+        self
+    }
+
+    //
+    // Add a formatted payload as a single argument to the flash program that
+    // consists of the prefix followed by the path to the payload, followed by
+    // the suffix.
+    //
+    fn formatted_payload<'a>(
+        &'a mut self,
+        prefix: &str,
+        suffix: &str,
+    ) -> &'a mut Self {
+        self.args.push(FlashArgument::FormattedPayload(
+            prefix.to_string(),
+            suffix.to_string(),
+        ));
+        self
+    }
+
+    //
+    // Add a flasher configuration file as an argument to the flash program
+    //
+    fn config<'a>(&'a mut self) -> &'a mut Self {
+        self.args.push(FlashArgument::Config);
+        self
+    }
+
+    //
+    // Slurp in any flash program configuration file and flatten it into
+    // our overall configuration
+    //
+    pub fn flatten(&mut self) -> anyhow::Result<()> {
+        if let FlashProgram::OpenOcd(ref config) = self.program {
+            if let FlashProgramConfig::Path(ref path) = config {
+                let p: PathBuf = path.iter().collect();
+                let text = std::fs::read_to_string(p)?;
+                self.program =
+                    FlashProgram::OpenOcd(FlashProgramConfig::Payload(text));
```

Vague architectural complaint:

`FlashProgramConfig::Path` is only ever used here in `run`; `FlashProgramConfig::Payload` is only ever written to a `flash.ron` file.  They're the same (enum) type, but having the wrong enum will break things either here or in Humility.

It would be less error-prone to _always_ flatten (and modify `run` below); then we could do flattening immediately in `FlashProgramConfig::new` and remove the `enum` entirely.

---

So, that would mean that in the case of flashing live, we are (unnecessarily) slurping in the file and then writing it back out again to a temporary file.  Happy to do that, but it feels like unnecessary complexity?  (Especially because the temporary file would be blown away by default.)

---

Yeah, I'm now on the fence after having read the Humility side of things (and noticing that it reads + writes a file).  I think this is okay for now, since it will fail loudly if you forget to flatten it.

Presumably, the future "shared Hubris + Humility functionality" crate would handle both cases 😄